staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go

```diff
@@ -151,9 +154,20 @@ func (f *FieldManager) Apply(liveObj runtime.Object, patch []byte, force bool) (
 	if err != nil {
 		return nil, fmt.Errorf("failed to decode managed fields: %v", err)
 	}
-	// We can assume that patchObj is already on the proper version:
-	// it shouldn't have to be converted so that it's not defaulted.
-	// TODO (jennybuckley): Explicitly checkt that patchObj is in the proper version.
+	// Check that the patch object has the same version as the live object
+	unstructuredPatch := &unstructured.Unstructured{Object: map[string]interface{}{}}
+
+	if err := yaml.Unmarshal(patch, &unstructuredPatch.Object); err != nil {
+		return nil, fmt.Errorf("error decoding YAML: %v", err)
+	}
+	if unstructuredPatch.GetAPIVersion() != f.groupVersion.String() {
+		return nil,
+			errors.NewBadRequest(
+				fmt.Sprintf("Incorrect version specified in apply patch. "+
+					"Specified patch version: %s, expected: %s",
+					unstructuredPatch.GetAPIVersion(), f.groupVersion.String()))
+	}
+
```

Comment:
Just thought of something, can you change line 177 below here from
` patchObjTyped, err := f.typeConverter.YAMLToTyped(patch) `
to
` patchObjTyped, err := f.typeConverter.ObjectToTyped(patchObj) `
so we don't have to do the unstructured de-serialization twice

Reply:
good idea! done

Reply:
Sorry, I've thought about this more and it turns out that I was wrong about this suggestion, because we need to create the TypedValue directly from YAML to preserve float/int distinction. The rest of this PR is great, I'll try to get the milestone added here

Reply:
For sure, I've changed it back to:
'patchObjTyped, err := f.typeConverter.YAMLToTyped(patch)'

Reply:
thanks!